### PR TITLE
[Utility][RPC][CLI] Querying governance parameters (Issue #619)

### LINKED ISSUE #619: [Utility] Querying governance params
## Objective

Improve the ability to access on-chain params while improving how they are maintained.

### V0 Parametesr

[The utility specification](https://github.com/pokt-network/pocket-network-protocol/tree/main/utility) mentions a non-exhaustive list of on-chain parameters that will be used in V1.

The V0 RPC spec can be found [here](https://editor.swagger.io/?url=https://raw.githubusercontent.com/pokt-network/pocket-core/staging/doc/specs/rpc-spec.yaml), allowing to query the v0 params like so:

```bash
curl --location -X 'POST' -H 'Content-Type: application/json' 'https://<user>:<pass>@<hostname>:<port>/v1/query/allParams' | jq

{
  "app_params": [
    {
      "param_key": "application/StabilityAdjustment",
      "param_value": "0"
    },
    {
      "param_key": "application/ParticipationRateOn",
      "param_value": "false"
    },
    {
      "param_key": "application/ApplicationStakeMinimum",
      "param_value": "1000000"
    },
    {
      "param_key": "application/BaseRelaysPerPOKT",
      "param_value": "200000"
    },
    {
      "param_key": "application/AppUnstakingTime",
      "param_value": "1814000000000000"
    },
    {
      "param_key": "application/MaxApplications",
      "param_value": "2295"
    },
    {
      "param_key": "application/MaximumChains",
      "param_value": "15"
    }
  ],
  "auth_params": [
    {
      "param_key": "auth/FeeMultipliers",
      "param_value": "{\"fee_multiplier\":null,\"default\":\"1\"}"
    },
    {
      "param_key": "auth/MaxMemoCharacters",
      "param_value": "75"
    },
    {
      "param_key": "auth/TxSigLimit",
      "param_value": "8"
    }
  ],
  "gov_params": [
    {
    ...
```

### V1 Paramters

The code to list and maintain governance parameters is spread throughout the codebase. A good starting point would be to look at all `*gov*` files:

Screenshot 2023-03-24 at 12 22 54 PM

For example, [gov_test.go](https://github.com/pokt-network/pocket/blob/main/utility/gov_test.go) or [gov.go](https://github.com/pokt-network/pocket/blob/main/utility/gov.go)

## Goals

- Enable querying localnet params via the CLI
- Reduce the code footprint of governance parameters. This has two underlying goles:
	- Easier to read & maintain code
	- Less context that will allow using this library with ChatGPT (more code -> more tokens -> less context)

## Deliverable

- [ ] A PR that reduces the code footprint of everything related to maintaining governance paramsSelect and document build tags used to include/exclude actor-specific business logic
- [ ] Add a `/v1/query/nodeParams` to the [v1 RPC spec](https://editor.swagger.io/?url=https://raw.githubusercontent.com/pokt-network/pocket/main/rpc/v1/openapi.yaml).
- [ ] A make target that helps querying these params on localnet

## Non-goals / Non-deliverables

- Adding/removing any existing parameters in the codebase

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **Required**: A make target that queries the on-chain parameters of localnet
- [ ] **All tests**: `make test_all`
- [ ] **LocalNet**: verify a `LocalNet` is still functioning correctly by following the instructions at [docs/development/README.md](https://github.com/pokt-network/pocket/tree/main/docs/development)

---

**Creator**: @Olshansk 
**Co-Owners**: @jessicadaugherty

### PR DESCRIPTION
**Pull request by h5law**
## Description

This PR introduces the ability to query `v1/query/allChainParmas` with a `GET` request in order to retrieve the current values for all governance parameters stored in the node's PostgreSQL database. This is defined in the OpenAPI YAML RPC specification and its logic is implemented in `rpc/handlers.go`

The CLI endpoint is a subcommand under the `System` command that allows for the querying of this RPC endpoint and is used by the make target `query_node_params` to quickly enable the querying of governance parameters on localnet.

In addition to this the `GovParamMetadatMap` has been moved to `shared/utils/gov_utils.go` where it is being initialised and used in `persistance` and `utility`. The `owner` tags have been added to the `genesis.proto` file to show which parameters own which, this allows for an easy lookup and reduces the code footprint of the `getParamOwner` function greatly.

There has also been a lot of code cleanup in `utility` simplifying governance parameter accessor functions and their tests

## Issue

Fixes #619 
## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Add `GetAllParams()` function to the persistence read interface to retrieve the current value of each parameter
- Add the `v1/query/allChainParams` RPC endpoint to the RPC specification, and implement its logic
- Add the `Query` subcommand in the CLI
- Add the `AllChainParams` subcommand to the Query CLI endpoint to query `v1/query/allChainParams`
- Add `make query_chain_params` target to easily query the parameters on a localnet
- Move parameter metadata logic into `shared/utils/gov.go`
- Reduce code footprint of gov parameters, their accessor functions and tests

## Testing

- [x] `make develop_test`
- [x] [LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md) w/ all of the steps outlined in the `README`

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [x] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: utility/unit_of_work/gov.go [hunk 1]
```diff
@@ -3,12 +3,108 @@ package unit_of_work
...
+
+const (
+	BIGINT int = iota
+	INT
+	INT64
+	BYTES
+	STRING
+)
+
+func generateParamConfigs() map[string]int {
```
Comment (Olshansk) [inline utility/unit_of_work/gov.go:29]: I know this is overly explicit, but I feel like do you think this helps? `prepareGovParamParamTypesMap`

File: utility/unit_of_work/gov.go [hunk 2]
```diff
@@ -3,12 +3,108 @@ package unit_of_work
...
 	typesUtil "github.com/pokt-network/pocket/utility/types"
 	"google.golang.org/protobuf/types/known/wrapperspb"
 )
 
+func init() {
+	GovParams = generateParamConfigs()
+}
+
+var (
+	GovParams map[string]int
```
Comment (Olshansk) [inline utility/unit_of_work/gov.go:18]: 1. Confirming that this needs to be exposed?
2. Maybe `GovParamTypes`?

File: shared/utils/gov_utils.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	"reflect"
+	"strings"
+
+	"github.com/pokt-network/pocket/runtime/genesis"
+)
+
+// init initializes a map that contains the metadata extracted from `gov.proto`.
+// Since protobuf files do not change at runtime, it seems efficient to do it here.
+func init() {
+	GovParamMetadataMap = parseGovProto()
```
Comment (Olshansk) [inline shared/utils/gov_utils.go:13]: Can we verify that `GovParams` (the map you have hardcoded elsewhere) corresponds to `gov.proto`?

I basically want a compile check to make sure we don't edit one file and forget to update the source code.

**Review by Olshansk — CHANGES_REQUESTED** [review]
This is 🪨 work. Looking solid

File: app/client/cli/system.go [hunk 4]
```diff
@@ -74,6 +76,33 @@ func systemCommands() []*cobra.Command {
...
+				}
+				statusCode := response.StatusCode
+				body, err := io.ReadAll(response.Body)
+				if err != nil {
+					logger.Global.Error().Err(err).Msg("Error reading response body")
+					return err
+				}
+				if statusCode == http.StatusOK {
+					fmt.Println(string(body))
+					return nil
```
[the remark this answers is not in the document]
Comment (h5law) [reply app/client/cli/system.go:101]: When we pipe to `jq` we are piping `stdout` so the `fmt.Println(string(body))` is what is being piped to `jq` I use the `make query_node_params | jq` command with localnet fine

File: persistence/gov.go [hunk 5]
```diff
@@ -215,11 +189,41 @@ func (p *PostgresContext) getFlagsUpdated(height int64) ([]*coreTypes.Flag, erro
...
+}
+
 func (p *PostgresContext) getParamsOrFlagsUpdateAtHeightQuery(tableName string, height int64) string {
 	fields := "name,value"
 	if tableName == types.FlagsTableName {
 		fields += ",enabled"
 	}
 	// Build correct query to get all Params/Flags at certain height ordered by their name values
-	return fmt.Sprintf(`SELECT %s FROM %s WHERE height=%d ORDER BY name ASC`, fields, tableName, height)
+	return fmt.Sprintf("SELECT %s FROM %s WHERE height=%d ORDER BY name ASC", fields, tableName, height)
```
[the remark this answers is not in the document]
Comment (h5law) [reply persistence/gov.go:219]: Here I actually just changed the ticks to match the quote marks, regarding the sort in 2 places - we have to sort in code as they are put into a map,  and maps don't keep their order. Here _I think_ we still need to keep the sort as the way I am getting the latest values is with `PostgreSQL` native `DISTICT` command. So the sorting by name ascending and height descending means we take only the most recent entry of each without any more complex filters

File: rpc/handlers.go [hunk 6]
```diff
@@ -50,6 +51,35 @@ func (s *rpcServer) GetV1ConsensusState(ctx echo.Context) error {
...
+	paramValueMap, err := persistenceRC.GetAllParams()
+	if err != nil {
+		return ctx.String(http.StatusInternalServerError, err.Error())
+	}
+	parameterKeyValues := make([]*paramValue, 0)
+	keys := make([]string, 0)
+	for key := range paramValueMap {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
```
[the remark this answers is not in the document]
Comment (h5law) [reply rpc/handlers.go:69]: So the reason I sort here Is because its a map, instead I will try to make a slice like this `{name, value},{name, value}` this might be more efficient

File: shared/utils/gov_utils.go (hunk 3 above)
Comment (h5law) [reply shared/utils/gov_utils.go:13]: I am unsure what you mean? I have removed the other maps that do that same thing to consolidate them here. This reflects the `genesis.Params` struct and doesn't need any updating if that changes.

File: shared/utils/gov_utils.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+	for _, field := range fields {
+		if !field.IsExported() {
+			continue
+		}
+		poktTag := field.Tag.Get("pokt")
+		protoTag := field.Tag.Get("protobuf")
+		poktValType := extractStructTag(poktTag, "val_type=")
+		poktOwner := extractStructTag(poktTag, "owner=")
+		golangType := field.Type.Name() // Get string version of field's Golang type
+		protoName := extractStructTag(protoTag, "name=")
```
[the remark this answers is not in the document]
Comment (h5law) [reply shared/utils/gov_utils.go:49]: This does exist, protobuf generates it, but we don't use this field so removed it 👍🏼

File: utility/unit_of_work/gov.go [hunk 8]
```diff
@@ -3,12 +3,108 @@ package unit_of_work
...
+		typesUtil.MessageStakeServicerFee:                  BIGINT,
+		typesUtil.MessageEditStakeServicerFee:              BIGINT,
+		typesUtil.MessageUnstakeServicerFee:                BIGINT,
+		typesUtil.MessagePauseServicerFee:                  BIGINT,
+		typesUtil.MessageUnpauseServicerFee:                BIGINT,
+		typesUtil.MessageChangeParameterFee:                BIGINT,
+	}
+}
+
+func getGovParam[T *big.Int | int | int64 | []byte | string](u *baseUtilityUnitOfWork, paramName string) (i T, err typesUtil.Error) {
```
[the remark this answers is not in the document]
Comment (h5law) [reply utility/unit_of_work/gov.go:85]: We cannot make it a reciever as methods in interfaces and of structs etc cannot have type parameters

File: utility/unit_of_work/gov.go [hunk 9]
```diff
@@ -33,498 +129,71 @@ func (u *baseUtilityUnitOfWork) updateParam(paramName string, value any) typesUt
 	return typesUtil.ErrUnknownParam(paramName)
 }
 
-func (u *baseUtilityUnitOfWork) getParameter(paramName string) (any, error) {
```
[the remark this answers is not in the document]
Comment (h5law) [reply utility/unit_of_work/gov.go:36]: My guy

File: utility/unit_of_work/gov_test.go [hunk 10]
```diff
@@ -13,498 +16,64 @@ import (
...
-	require.NoError(t, err)
-	require.Equal(t, defaultParam, gotParam)
+func TestUtilityUnitOfWork_GetGovParams(t *testing.T) {
+	uow := newTestingUtilityUnitOfWork(t, 0)
+	defaultParams := DefaultTestingParams(t)
+	for _, paramName := range utils.GovParamMetadataKeys {
+		if strings.Contains(paramName, "_owner") {
+			continue
+		}
+		defaultParam := reflect.ValueOf(defaultParams).MethodByName("Get" + utils.GovParamMetadataMap[paramName].PropertyName).Call([]reflect.Value{})[0].Interface()
```
[the remark this answers is not in the document]
Comment (h5law) [reply utility/unit_of_work/gov_test.go:30]: Yes I will rename

File: persistence/test/gov_test.go [hunk 11]
```diff
@@ -189,5 +189,26 @@ func TestGetSetToggleByteArrayFlag(t *testing.T) {
...
 	require.Equal(t, true, enabled)
+}
+
+func TestGetAllParams(t *testing.T) {
+	db := NewTestPostgresContext(t, 0)
 
+	err := db.InitGenesisParams(test_artifacts.DefaultParams())
+	require.NoError(t, err)
+
+	appMaxChain, err := db.GetStringParam(AppMaxChainsParamName, 0)
```
[the remark this answers is not in the document]
Comment (h5law) [reply persistence/test/gov_test.go:200]: I think this test involves setting parameters too, I think its best to keep these 3 tests separate as I have opted to keep the `GetStringParam` `GetBytesParam` and `GetIntParam` functions alive so that we can use the generic wrapper `GetParameter` now which is not an interface function.

**Comment by h5law** [comment]
[Screencast from 2023-04-07 02-09-59.webm](https://user-images.githubusercontent.com/53987565/230520344-5823caf5-bf9c-4ac2-a873-ec6b24bc7d92.webm)

Just to confirm this does work with `jq`

File: utility/unit_of_work/block_test.go [hunk 12]
```diff
@@ -124,10 +124,10 @@ func TestUtilityUnitOfWork_EndBlock(t *testing.T) {
...
 
-	feeBig, err := uow.getMessageSendFee()
+	feeBig, err := getGovParam[*big.Int](uow, utilTypes.MessageSendFee)
 	require.NoError(t, err)
 
-	proposerCutPercentage, err := uow.getProposerPercentageOfFees()
+	proposerCutPercentage, err := getGovParam[int](uow, utilTypes.ProposerPercentageOfFeesParamName)
 	require.NoError(t, err)
 
 	feesAndRewardsCollectedFloat := new(big.Float).SetInt(feeBig)
```
[the remark this answers is not in the document]
Comment (h5law) [reply utility/unit_of_work/block_test.go:133]: This is not correct as we are not type casting we are using type parameters as part of the function signature

File: utility/unit_of_work/actor_test.go [hunk 13]
```diff
@@ -263,13 +263,13 @@ func TestUtilityUnitOfWork_GetUnbondingHeight(t *testing.T) {
...
+				unstakingBlocks, err = getGovParam[int64](uow, typesUtil.FishermanUnstakingBlocksParamName)
 			case coreTypes.ActorType_ACTOR_TYPE_SERVICER:
-				unstakingBlocks, err = uow.getServicerUnstakingBlocks()
+				unstakingBlocks, err = getGovParam[int64](uow, typesUtil.ServicerUnstakingBlocksParamName)
 			case coreTypes.ActorType_ACTOR_TYPE_VAL:
-				unstakingBlocks, err = uow.getValidatorUnstakingBlocks()
+				unstakingBlocks, err = getGovParam[int64](uow, typesUtil.ValidatorUnstakingBlocksParamName)
 			default:
 				t.Fatalf("unexpected actor type %s", actorType.String())
 			}
```
[the remark this answers is not in the document]
Comment (h5law) [reply utility/unit_of_work/actor_test.go:275]: 1 & 2 ignore
3 - I dont think this is necessary.
4 - Error is handled with `require.NoError(t, err)` after the switch etc

File: shared/modules/persistence_module.go [hunk 14]
```diff
@@ -196,7 +196,7 @@ type PersistenceReadContext interface {
 	GetIntParam(paramName string, height int64) (int, error)
 	GetStringParam(paramName string, height int64) (string, error)
 	GetBytesParam(paramName string, height int64) ([]byte, error)
-	GetParameter(paramName string, height int64) (any, error)
+	GetAllParams() ([][]string, error)
 
 	// Flags
 	GetIntFlag(paramName string, height int64) (int, bool, error)
```
[the remark this answers is not in the document]
Comment (h5law) [reply shared/modules/persistence_module.go:202]: So we use a slice of slices instead of a map to maintain the sorted order from the DB query, this allows us to speed up returning the key-value pairs by not having to sort them again (as maps don't maintain order) when we load them into the `GetV1QueryAllChainParams` response.

I don't want to bring back `GetParameter` as an interface function as the function using type parameters is a lot nicer in my opinion Mr. AI

File: utility/unit_of_work/gov_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply utility/unit_of_work/gov_test.go:30]: Can you rename it to `defaultValueParam` - it'll make it a bit clearer.

File: utility/unit_of_work/gov.go [hunk 15]
```diff
@@ -3,12 +3,108 @@ package unit_of_work
...
+
+const (
+	BIGINT int = iota
+	INT
+	INT64
+	BYTES
+	STRING
+)
+
+func prepareGovParamParamTypesMap() map[string]int {
```
Comment (Olshansk) [inline utility/unit_of_work/gov.go:29]: Regarding this comment: https://github.com/pokt-network/pocket/pull/622/files/010c045d060cf8114d09066846c1f7d1fe5f7318..2158a38a6ff9529a405a23c55366e37d403afa5b#r1160332826

If we update `genesis.proto`, there's nothing currently that forces us to update `prepareGovParamParamTypesMap` as well.

Source of truth: `genesis.proto`
Derived source of truth based on above: `GovParamMetadataMap`
Alternate source of truth: `prepareGovParamParamTypesMap`

In the forloop, we currently ignore missing ids: `switch id := govParamTypes[paramName]; id {`

tl;dr I basically want a fatal error if the ID doesn't exist.

File: shared/utils/gov_utils.go (hunk 3 above)
Comment (Olshansk) [reply shared/utils/gov_utils.go:13]: Leaving a comment in `utility/unit_of_work/gov.go`

File: shared/modules/persistence_module.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply shared/modules/persistence_module.go:202]: Show that AI who's boss

File: utility/unit_of_work/actor_test.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply utility/unit_of_work/actor_test.go:275]: Not valuable, but it's still interesting to learn where AI stands with regard to coding reviews.

I'll give it a couple years before it can replace us 😅

File: utility/unit_of_work/gov.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply utility/unit_of_work/gov.go]: Got it. DIdn't even know that.

Let's leave it as is. I was looking for solutions and it seems like building a wrapper (e.g. https://rakyll.org/generics-facilititators/) is the most common approach, but that's overkill for what we need.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Just a few more minor nits and then good2go

File: utility/unit_of_work/gov.go (hunk 15 above)
Comment (h5law) [reply utility/unit_of_work/gov.go:29]: Well this will always result in an error as the main map has the owners in it too, but this type map doesn't. I will add some logic to the init where it loops over the non-owner keys and checks they have a type in the map

Comment (h5law) [reply utility/unit_of_work/gov.go:29]: Good think we did this I missed 2 😅

Comment (Olshansk) [reply utility/unit_of_work/gov.go:39]: Paying dividends on day 0.

**Review by Olshansk — APPROVED** [review]
👨‍⚖️ 👍
